test/Microsoft.ML.Tests/TrainerEstimators/TreeEnsembleFeaturizerTest.cs

```diff
@@ -108,5 +115,633 @@ public void TreeEnsembleFeaturizerOutputSchemaTest()
+            var predicted = treeModel.Transform(dataView);
+
+            // Combine the output of TreeEnsembleFeaturizationTransformer and the original features as the final training features.
+            // Then train a linear model.
+            var options = new PretrainedTreeFeaturizationEstimator.Options()
+            {
+                InputColumnName = "Features",
+                TreesColumnName = "Trees",
+                LeavesColumnName = "Leaves",
+                PathsColumnName = "Paths",
+                ModelParameters = treeModel.Model.SubModel
+            };
+            var pipeline = ML.Transforms.FeaturizeByPretrainTreeEnsemble(options).
+                Append(ML.Transforms.Concatenate("CombinedFeatures", "Features", "Trees", "Leaves", "Paths")).
+                Append(ML.BinaryClassification.Trainers.SdcaLogisticRegression("Label", "CombinedFeatures"));
+            var model = pipeline.Fit(dataView);
+            var prediction = model.Transform(dataView);
+            var metrics = ML.BinaryClassification.Evaluate(prediction);
+
+            // Then train the same linear model without tree features.
+            var naivePipeline = ML.BinaryClassification.Trainers.SdcaLogisticRegression("Label", "Features");
+            var naiveModel = naivePipeline.Fit(dataView);
+            var naivePrediction = naiveModel.Transform(dataView);
+            var naiveMetrics = ML.BinaryClassification.Evaluate(naivePrediction);
+
+            // The linear model trained with tree features should perform better than that without tree features.
+            Assert.True(metrics.Accuracy > naiveMetrics.Accuracy);
+            Assert.True(metrics.LogLoss < naiveMetrics.LogLoss);
+            Assert.True(metrics.AreaUnderPrecisionRecallCurve > naiveMetrics.AreaUnderPrecisionRecallCurve);
+        }
+
+        [Fact]
+        public void TestFastTreeBinaryFeaturizationInPipeline()
+        {
+            int dataPointCount = 200;
+            var data = SamplesUtils.DatasetUtils.GenerateBinaryLabelFloatFeatureVectorFloatWeightSamples(dataPointCount).ToList();
+            var dataView = ML.Data.LoadFromEnumerable(data);
+
+            var trainerOptions = new FastTreeBinaryTrainer.Options
+            {
+                    NumberOfThreads = 1,
+                    NumberOfTrees = 10,
+                    NumberOfLeaves = 4,
+                    MinimumExampleCountPerLeaf = 10,
+                    FeatureColumnName = "Features",
+                    LabelColumnName = "Label"
+            };
+
+            var options = new FastTreeBinaryFeaturizationEstimator.Options()
+            {
+                InputColumnName = "Features",
+                TreesColumnName = "Trees",
+                LeavesColumnName = "Leaves",
+                PathsColumnName = "Paths",
+                TrainerOptions = trainerOptions
+            };
+
+            var pipeline = ML.Transforms.FeaturizeByFastTreeBinary(options).
```

Comment:
Creating (8) seperate `ML.Transforms.FeaturizeBy{ FastTreeBinary, FastForestRegression, FastTreeRegression, FastTreeTweedie, ... }` featurizers under `ML.Transforms.*` seem a bit large. Seems to clutter up the namespace. 

In the future, this should grow as I think we should have LightGBM variants too (and Catboost if we take it in). The number of independent featurizers will be: 
{ FastTree, FastTreeTweedie, FastForest, LightGBM, CatBoost } x { BinaryClassification, Multiclass, Regression, Ranking }`. _(with some combinations missing)_

Would it be more clean to have one `ML.Transforms.TreeFeaturizer()`, and put the specific instance type as a parameter? Would it be doable to have one return type?

Reply:
There are two reasons that I don't like a single `TreeFeaturizer`.

1.  It goes an opposite direction of the C# API's design. Most of them are strongly typed to the underlying data structures. You can see we have a lot of `FastTree...` and `LightGbm...`.
2. It may requires user to specify `TreeFeaturizer<TTrainer>(options)` and the user manually need to make sure the type of `options` is `TTrainer.Options`, which is easy to make mistakes.

Reply:
You're right, the current pattern will likely have less mistakes by the user. 

For the plethora of `FastTree...` and `LightGBM...`, those are namespaced under the task `mlContext.Regression.Trainers.FastTree()`, hence the duplication isn't visible.